File: legographtest.js
"use strict";

const cas = require("cassowary");
const helpers = require("./helpers");
const G = require("./model");

const addLinkToCSolver = (solver, nBrickA, nStudA, nBrickB, nHoleB) => {
    const studPos = G.bricks[nBrickA].worldStuds[nStudA];
    const holePos = G.bricks[nBrickB].worldHoles[nHoleB];
    helpers.equaliseTwoArrays(solver, studPos, holePos);
};

// refactor to reduce repititiveness
const addWorldRelationToCSolver = (solver, nBrick) => {
    const brick = G.bricks[nBrick];
    const brickPos = brick.pos;

    // process stud A and stud B
    for (let j = 0 ; j < 2 ; j+=1) {
        // stud j's X, Y, Z co-ordinates
        for (let i = 0 ; i < 3 ; i+=1) {
            solver.addConstraint(new cas.Equation(brick.worldStuds[j][i],
                                                  cas.plus(brick.localStuds[j][i], brickPos[i])));
            solver.addConstraint(new cas.Equation(brick.worldHoles[j][i],
                                                  cas.plus(brick.localHoles[j][i], brickPos[i])));
        }
    }
}

const testConstraintSolver = () => {
    const solver = new cas.SimplexSolver();

    addWorldRelationToCSolver(solver, 0);
    addWorldRelationToCSolver(solver, 1);
    addWorldRelationToCSolver(solver, 2);

    addLinkToCSolver(solver, 0, 1, 1, 0);
    addLinkToCSolver(solver, 1, 0, 2, 0);
    addLinkToCSolver(solver, 1, 1, 2, 1);

    solver.resolve();

    console.log("Brick 0 position : ", G.bricks[0].pos);
    console.log("Brick 1 position : ", G.bricks[1].pos);
    console.log("Brick 2 position : ", G.bricks[2].pos);
};
testConstraintSolver();

File: lego_graphtest.py
import numpy as np
import scipy.optimize
import helpers as h
import model as m

relations = []

def addLinkToSolver(nBrickA, nStudA, nBrickB, nHoleB):
    studPos = m.bricks[nBrickA].worldStuds[nStudA]
    holePos = m.bricks[nBrickB].worldHoles[nHoleB]
    h.equaliseTwoLists(studPos, holePos)

def addWorldRelationToSolver(relations, nBrick):
    brick = m.bricks[nBrick]
    brickPos = brick.pos
    for j in range(0, 2):
        worldStud = brick.worldStuds[j]
        worldHole = brick.worldHoles[j]
        localStud = brick.localStuds[j]
        localHole = brick.localHoles[j]
        for i in range(0, 3):
            fn_str = 'lambda V: ' + worldStud[i].to_string() + ' - ' + localStud[i].to_string() + ' - ' + brickPos[i]
            fn = eval(fn_str)
            relations.append(fn)

def F(V):
    return [f(V) for f in relations]

def testConstraintSolver():
    addWorldRelationToSolver(relations, 0)
    addWorldRelationToSolver(relations, 1)
    addWorldRelationToSolver(relations, 2)

    addLinkToSolver(0, 1, 1, 0)
    addLinkToSolver(1, 0, 2, 0)
    addLinkToSolver(1, 1, 2, 1)

    # put first brick into [0, 0, 0]
    for i in range(0, 3):
        fn_str = 'lambda V: ' + m.bricks[0].pos[i].to_string()
        fn = eval(fn_str)
        relations.append(fn)
    V = scipy.optimize.broyden1(F, [0] * h.counter, f_tol=1e-14)
    print(V)

testConstraintSolver()
